utilities/transactions/pessimistic_transaction_db.cc

```diff
@@ -499,5 +503,160 @@ void PessimisticTransactionDB::UnregisterTransaction(Transaction* txn) {
+  // commit_seq lower than any live snapshot, including snapshot_seq.
+  if (old_commit_map_empty_.load(std::memory_order_acquire)) {
+    return true;
+  }
+  {
+    // We should not normally reach here
+    ReadLock rl(&old_commit_map_mutex_);
+    auto old_commit_entry = old_commit_map_.find(prep_seq);
+    if (old_commit_entry == old_commit_map_.end() ||
+        old_commit_entry->second <= snapshot_seq) {
+      return true;
+    }
+  }
+  // (ii) it the case: it is committed but after the snapshot_seq
+  return false;
+}
+
+void WritePreparedTxnDB::AddPrepared(uint64_t seq) { prepared_txns_.push(seq); }
+
+void WritePreparedTxnDB::AddCommitted(uint64_t prepare_seq,
+                                      uint64_t commit_seq) {
+  auto indexed_seq = prepare_seq % COMMIT_CACHE_SIZE;
+  uint64_t evicted_prep_seq, evicted_commit_seq;
+  bool to_be_evicted =
+      GetCommitEntry(indexed_seq, &evicted_prep_seq, &evicted_commit_seq);
+  if (to_be_evicted) {
+    if (max_evicted_seq_.load(std::memory_order_acquire) < evicted_commit_seq) {
+      auto max_evicted_seq = evicted_commit_seq;
+      max_evicted_seq_.store(max_evicted_seq, std::memory_order_release);
+      // When max_evicted_seq_ advances, move older entries from prepared_txns_
+      // to delayed_prepared_. This guarantees that if a seq is lower than max,
+      // then it is not in prepared_txns_ ans save an expensive, synchronized
+      // lookup from a shared set. delayed_prepared_ is expected to be empty in
+      // normal cases.
+      {
+        WriteLock wl(&prepared_mutex_);
+        while (!prepared_txns_.empty() &&
+               prepared_txns_.top() > max_evicted_seq) {
+          auto to_be_popped = prepared_txns_.top();
+          delayed_prepared_.insert(to_be_popped);
+          prepared_txns_.pop();
+          delayed_prepared_empty_.store(false, std::memory_order_release);
+        }
+      }
+      {
+        WriteLock wl(&snapshots_mutex_);
+        InstrumentedMutex(db_impl_->mutex());
+        snapshots_ = db_impl_->snapshots().GetAll();
+      }
+    }
+    // After each eviction from commit cache, check if the commit entry should
+    // be kept around because it overlaps with a live snapshot.
+    {
+      ReadLock rl(&snapshots_mutex_);
+      for (auto snapshot : snapshots_) {
+        auto snapshot_seq =
+            reinterpret_cast<const SnapshotImpl*>(snapshot)->number_;
+        if (evicted_commit_seq <= snapshot_seq) {
+          break;
+        }
+        // then snapshot_seq < evicted_commit_seq
+        if (evicted_prep_seq <= snapshot_seq) {  // overlapping range
+          old_commit_map_empty_.store(false, std::memory_order_release);
+          WriteLock wl(&old_commit_map_mutex_);
+          old_commit_map_[evicted_prep_seq] = evicted_commit_seq;
+        }
+      }
+    }
+  }
+  AddCommitEntry(indexed_seq, prepare_seq, commit_seq);
+  {
+    WriteLock wl(&prepared_mutex_);
+    prepared_txns_.erase(prepare_seq);
```

Comment:
You'll put it in `erased_prepared_txns_` on erase and never correctly pop the value. You need to fix that.